### PR TITLE
[Issue 4283][flink] construct auth when building pulsar source

### LINKED ISSUE #4283: flink job crashes when serializing auth in pulsar client
**Describe the bug**
pulsar flink connector now uses ClientConfigData to instantiate pulsar client. Pulsar client composes Authentication which can not be serialized. In an environment with Auth there is no way to set auth in pulsar client.

**To Reproduce**
The below code crashes on last line:

>         ClientConfigurationData clientConfig = new ClientConfigurationData();
>         clientConfig.setServiceUrl(PULSAR_BROKER_URL);
>         clientConfig.setAuthPluginClassName(AuthenticationToken.class.getName());
>         clientConfig.setAuthParams(ImmutableMap.of("token", PULSAR_AUTH_TOKEN));
> 
>         ConsumerConfigurationData consumerConfig = new ConsumerConfigurationData();
>         consumerConfig.setSubscriptionInitialPosition(SubscriptionInitialPosition.Earliest);
>         consumerConfig.setTopicNames(new HashSet<>(Arrays.asList(PULSAR_TOPIC)));
>         consumerConfig.setSubscriptionName(PULSAR_SUBSCRIPTION_NAME);
> 
>         PulsarSourceBuilder pulsarSourceBuilder = PulsarSourceBuilder.builder(new SimpleStringSchema())
>                 .pulsarAllClientConf(clientConfig)
>                 .pulsarAllConsumerConf(consumerConfig);
> 
>         StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
> 
>         env.setStreamTimeCharacteristic(TimeCharacteristic.ProcessingTime);
> 
>         SourceFunction src = pulsarSourceBuilder.build();
>         DataStream dataStream = env.addSource(src).uid("testStream");

**Expected behavior**
Provide a way to set Auth for pulsar client from pulsar flink connector

### PR DESCRIPTION
**Pull request by shiv4289**
<--
### Contribution Checklist
 
Fixes #4283

### Motivation

pulsar flink connector now uses ClientConfigData to instantiate pulsar client. Pulsar client composes Authentication which can not be serialized. In an environment with Auth there is no way to set auth in pulsar client.

### Modifications

Keep auth params away from persistence and serialization. Construct auth when building pulsar source

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API: (yes)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no )

### Documentation
NA

### CODE REVIEW

**Comment by shiv4289** [comment]
run integration tests
run java8 tests

File: pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/PulsarSourceBuilder.java [hunk 1]
```diff
@@ -277,9 +278,27 @@ private PulsarSourceBuilder(DeserializationSchema<T> deserializationSchema) {
...
         return new PulsarConsumerSource<>(this);
     }
 
+    private void setTransientFields() throws PulsarClientException {
+        setAuth();
+    }
+
+    private void setAuth() throws PulsarClientException{
+        if (StringUtils.isBlank(this.clientConfigurationData.getAuthPluginClassName())
+                && this.clientConfigurationData.getAuthParams() == null || this.clientConfigurationData.getAuthParams().isEmpty())
```
Comment (merlimat) [inline pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/PulsarSourceBuilder.java]:
```suggestion
                && StringUtils.isBlank(this.clientConfigurationData.getAuthParams()))
```

Comment (shiv4289) [reply pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/PulsarSourceBuilder.java]: @merlimat this returns a map so thats not possible :-)

> this.clientConfigurationData.getAuthParams()

**Comment by shiv4289** [comment]
run integration tests
run java8 tests

File: pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/PulsarSourceBuilder.java (hunk 1 above)
Comment (merlimat) [reply pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/PulsarSourceBuilder.java]: Got it 👍

**Comment by shiv4289** [comment]
run integration tests
run java8 tests

**Comment by shiv4289** [comment]
run java8 tests

**Comment by shiv4289** [comment]
run integration tests

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java [hunk 2]
```diff
@@ -22,6 +22,7 @@
 import com.google.common.base.Charsets;
 
 import java.io.IOException;
+import java.io.Serializable;
```
Comment (sijie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java]: what is this import for?

Comment (shiv4289) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java]: Good catch, removed!

**Comment by merlimat** [comment]
run java8 tests

**Comment by shiv4289** [comment]
run java8 tests

**Pull request merged**